File: viewer/src/main/scala/farjs/viewer/quickview/QuickViewFile.scala
package farjs.viewer.quickview

import farjs.viewer._
import scommons.react._
import scommons.react.blessed._
import scommons.react.hooks._

import scala.scalajs.js

object QuickViewFile extends FunctionComponent[QuickViewFileProps] {

  private[quickview] var viewerController: ReactClass = ViewerController

  protected def render(compProps: Props): ReactElement = {
    val props = compProps.plain
    val (viewport, setViewport) = useState(js.undefined: js.UndefOr[ViewerFileViewport])

    val onKeypress: js.Function1[String, Boolean] = { key =>
      var processed = true
      key match {
        case "f3" =>
          props.inputRef.current.asInstanceOf[BlessedElement].emit("keypress", js.undefined, js.Dynamic.literal(
            name = "",
            full =
              if (props.isRight) ViewerEvent.onViewerOpenLeft
              else ViewerEvent.onViewerOpenRight
          ))
        case _ => processed = false
      }
      processed
    }

    <(viewerController)(^.plain := ViewerControllerProps(
      inputRef = props.inputRef,
      dispatch = props.dispatch,
      filePath = props.filePath,
      size = props.size,
      viewport = viewport,
      setViewport = setViewport,
      onKeypress = onKeypress
    ))()
  }
}

File: viewer/quickview/QuickViewFile.mjs
/**
 * @typedef {import("@farjs/blessed").Widgets.BlessedElement} BlessedElement
 * @import { Dispatch } from "@farjs/filelist/FileListData.mjs"
 * @import { ViewerFileViewport } from "../ViewerFileViewport.mjs";
 */
import React, { useState } from "react";
import ViewerController from "../ViewerController.mjs";
import ViewerEvent from "../ViewerEvent.mjs";

const h = React.createElement;

/**
 * @typedef {{
 *  readonly dispatch: Dispatch;
 *  readonly inputRef: React.MutableRefObject<BlessedElement | null>;
 *  readonly isRight: boolean;
 *  readonly filePath: string;
 *  readonly size: number;
 * }} QuickViewFileProps
 */

/**
 * @param {QuickViewFileProps} props
 */
const QuickViewFile = (props) => {
  const { viewerController } = QuickViewFile;

  const [viewport, setViewport] = useState(
    /** @type {ViewerFileViewport | undefined} */ (undefined)
  );

  /** @type {(key: string) => boolean} */
  const onKeypress = (key) => {
    let processed = true;
    const inputEl = props.inputRef.current;

    switch (key) {
      case "f3":
        inputEl?.emit("keypress", undefined, {
          name: "",
          full: props.isRight
            ? ViewerEvent.onViewerOpenLeft
            : ViewerEvent.onViewerOpenRight,
        });
        break;
      default:
        processed = false;
        break;
    }

    return processed;
  };

  return h(viewerController, {
    inputRef: props.inputRef,
    dispatch: props.dispatch,
    filePath: props.filePath,
    size: props.size,
    viewport,
    setViewport,
    onKeypress,
  });
};

QuickViewFile.displayName = "QuickViewFile";
QuickViewFile.viewerController = ViewerController;

export default QuickViewFile;
